# Lab | Data Structuring and Combining Data

## Challenge 1: Combining & Cleaning Data

In this challenge, we will be working with the customer data from an insurance company, as we did in the two previous labs. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

But this time, we got new data, which can be found in the following 2 CSV files located at the links below.

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv

Note that you'll need to clean and format the new data.

Observation:
- One option is to first combine the three datasets and then apply the cleaning function to the new combined dataset
- Another option would be to read the clean file you saved in the previous lab, and just clean the two new files and concatenate the three clean datasets

In [1]:
import pandas as pd

df1 = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv")
df2 = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv")
df3 = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv")

In [2]:
print(df1.columns)
print(df2.columns)
print(df3.columns)

Index(['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value',
       'Income', 'Monthly Premium Auto', 'Number of Open Complaints',
       'Policy Type', 'Vehicle Class', 'Total Claim Amount'],
      dtype='str')
Index(['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value',
       'Income', 'Monthly Premium Auto', 'Number of Open Complaints',
       'Total Claim Amount', 'Policy Type', 'Vehicle Class'],
      dtype='str')
Index(['Customer', 'State', 'Customer Lifetime Value', 'Education', 'Gender',
       'Income', 'Monthly Premium Auto', 'Number of Open Complaints',
       'Policy Type', 'Total Claim Amount', 'Vehicle Class'],
      dtype='str')


In [32]:
print(df1.head())
print(df2.head())
print(df3.head())

  customer       state gender             education customer_lifetime_value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    income  monthly_premium_auto number_of_open_complaints     policy_type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   vehicle_class  total_claim_amount  
0  Four-Door Car 

In [3]:
df1.columns = df1.columns.str.lower().str.replace(" ", "_")
df2.columns = df2.columns.str.lower().str.replace(" ", "_")
df3.columns = df3.columns.str.lower().str.replace(" ", "_")

In [4]:
df1 = df1.rename(columns={"st": "state"})
df2 = df2.rename(columns={"st": "state"})
# df3 zaten state, ona dokunmaya gerek yok

In [5]:
df1.isnull().sum()

customer                     2937
state                        2937
gender                       3054
education                    2937
customer_lifetime_value      2940
income                       2937
monthly_premium_auto         2937
number_of_open_complaints    2937
policy_type                  2937
vehicle_class                2937
total_claim_amount           2937
dtype: int64

In [6]:
df2.isnull().sum()

customer                     0
state                        0
gender                       5
education                    0
customer_lifetime_value      4
income                       0
monthly_premium_auto         0
number_of_open_complaints    0
total_claim_amount           0
policy_type                  0
vehicle_class                0
dtype: int64

In [7]:
df3.isnull().sum()

customer                     0
state                        0
customer_lifetime_value      0
education                    0
gender                       0
income                       0
monthly_premium_auto         0
number_of_open_complaints    0
policy_type                  0
total_claim_amount           0
vehicle_class                0
dtype: int64

In [8]:
df = pd.concat([df1, df2, df3])
df = df.reset_index(drop=True)
df.shape

(12074, 11)

In [9]:
df.columns

Index(['customer', 'state', 'gender', 'education', 'customer_lifetime_value',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'policy_type', 'vehicle_class', 'total_claim_amount'],
      dtype='str')

In [10]:
df["gender"] = df["gender"].replace({"Femal": "F", "female": "F", "Male": "M"})


In [11]:
df["education"] = df["education"].replace({"Bachelors": "Bachelor"})

In [12]:
def fix_clv(value):
    if isinstance(value, str):
        return value.replace("%", "")
    else:                               
        return value              

In [13]:
df["customer_lifetime_value"] = df["customer_lifetime_value"].apply(fix_clv) 

In [14]:
df["customer_lifetime_value"] = df["customer_lifetime_value"].astype(float)

In [15]:
df["vehicle_class"] = df["vehicle_class"].replace({"Sports Car": "Luxury", "Luxury SUV": "Luxury", "Luxury Car": "Luxury"})

In [16]:
def fix_complaints(value):
    if "/" in str(value):
        return str(value).split("/")[1]
    else:
        return value

In [17]:
df["number_of_open_complaints"] = df["number_of_open_complaints"].apply(fix_complaints)

In [18]:
df["number_of_open_complaints"] = df["number_of_open_complaints"].astype(float)

In [19]:
df = df.dropna(how="all")

In [20]:
df = df.drop_duplicates().reset_index(drop=True)

In [21]:
df["gender"].unique()

<StringArray>
[nan, 'F', 'M']
Length: 3, dtype: str

In [22]:
df["number_of_open_complaints"].unique()

array([0., 2., 1., 3., 5., 4.])

In [23]:
df["number_of_open_complaints"].dtype

dtype('float64')

In [24]:
df.isnull().sum()

customer                       0
state                          0
gender                       122
education                      0
customer_lifetime_value        7
income                         0
monthly_premium_auto           0
number_of_open_complaints      0
policy_type                    0
vehicle_class                  0
total_claim_amount             0
dtype: int64

In [25]:
df["customer_lifetime_value"].unique()

array([           nan,  697953.59    , 1288743.17    , ...,
          8163.890428,    7524.442436,    2611.836866])

In [26]:
test = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv")
test["Number of Open Complaints"].unique()

array([0, 2, 3, 1, 5, 4])

In [27]:
test["Customer Lifetime Value"].dtype

dtype('float64')

In [28]:
test["Customer Lifetime Value"].head

<bound method NDFrame.head of 0        3479.137523
1        2502.637401
2        3265.156348
3        4455.843406
4        7704.958480
            ...     
7065    23405.987980
7066     3096.511217
7067     8163.890428
7068     7524.442436
7069     2611.836866
Name: Customer Lifetime Value, Length: 7070, dtype: float64>

In [30]:
df["gender"] = df["gender"].fillna(df["gender"].mode()[0])
df["customer_lifetime_value"] = df["customer_lifetime_value"].fillna(df["customer_lifetime_value"].median())

In [34]:
print("df1'de % var mı?", df1["customer_lifetime_value"].astype(str).str.contains("%").any())
print("df2'de % var mı?", df2["customer_lifetime_value"].astype(str).str.contains("%").any())
print("df3'de % var mı?", df3["customer_lifetime_value"].astype(str).str.contains("%").any())

df1'de % var mı? True
df2'de % var mı? True
df3'de % var mı? False


In [31]:
df.isnull().sum()

customer                     0
state                        0
gender                       0
education                    0
customer_lifetime_value      0
income                       0
monthly_premium_auto         0
number_of_open_complaints    0
policy_type                  0
vehicle_class                0
total_claim_amount           0
dtype: int64

# Challenge 2: Structuring Data

In this challenge, we will continue to work with customer data from an insurance company, but we will use a dataset with more columns, called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by performing data cleaning, formatting, and structuring.

1. You work at the marketing department and you want to know which sales channel brought the most sales in terms of total revenue. Using pivot, create a summary table showing the total revenue for each sales channel (branch, call center, web, and mail).
Round the total revenue to 2 decimal points.  Analyze the resulting table to draw insights.

2. Create a pivot table that shows the average customer lifetime value per gender and education level. Analyze the resulting table to draw insights.

In [43]:
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv")
df.head()

,unnamed:_0,customer,state,customer_lifetime_value,response,coverage,education,effective_to_date,employmentstatus,gender,...,number_of_policies,policy_type,policy,renew_offer_type,sales_channel,total_claim_amount,vehicle_class,vehicle_size,vehicle_type,month
0,0,DK49336,Arizona,4809.216960,No,Basic,College,2011-02-18,Employed,M,...,9,Corporate Auto,Corporate L3,Offer3,Agent,292.800000,Four-Door Car,Medsize,A,2
1,1,KX64629,California,2228.525238,No,Basic,College,2011-01-18,Unemployed,F,...,1,Personal Auto,Personal L3,Offer4,Call Center,744.924331,Four-Door Car,Medsize,A,1
2,2,LZ68649,Washington,14947.917300,No,Basic,Bachelor,2011-02-10,Employed,M,...,2,Personal Auto,Personal L3,Offer3,Call Center,480.000000,SUV,Medsize,A,2
3,3,XL78013,Oregon,22332.439460,Yes,Extended,College,2011-01-11,Employed,M,...,2,Corporate Auto,Corporate L3,Offer2,Branch,484.013411,Four-Door Car,Medsize,A,1
4,4,QA50777,Oregon,9025.067525,No,Premium,Bachelor,2011-01-17,Medical Leave,F,...,7,Personal Auto,Personal L2,Offer1,Branch,707.925645,Four-Door Car,Medsize,A,1


In [44]:
marketing.columns

Index(['unnamed:_0', 'customer', 'state', 'customer_lifetime_value',
       'response', 'coverage', 'education', 'effective_to_date',
       'employmentstatus', 'gender', 'income', 'location_code',
       'marital_status', 'monthly_premium_auto', 'months_since_last_claim',
       'months_since_policy_inception', 'number_of_open_complaints',
       'number_of_policies', 'policy_type', 'policy', 'renew_offer_type',
       'sales_channel', 'total_claim_amount', 'vehicle_class', 'vehicle_size',
       'vehicle_type', 'month'],
      dtype='str')

In [49]:
df.pivot_table(index="sales_channel", values="total_claim_amount", aggfunc="sum").round(2)

,total_claim_amount
sales_channel,
Agent,1810226.82
Branch,1301204.00
Call Center,926600.82
Web,706600.04


The Agent channel generated the highest total revenue (1,810,226.82), nearly double that of the Web channel (706,600.04). This suggests that agents are the most effective sales channel, while web sales bring in the least revenue. The company might consider investing more in agent-based sales or improving the web channel's performance.

In [56]:
df.pivot_table(index="gender", columns="education", values="customer_lifetime_value", aggfunc="mean").round(2)

education,Bachelor,College,Doctor,High School or Below,Master
gender,,,,,
F,7874.27,7748.82,7328.51,8675.22,8157.05
M,7703.60,8052.46,7415.33,8149.69,8168.83


Customers with 'High School or Below' education have the highest average CLV for both genders (~8675 for F, ~8150 for M), while 'Doctor' education has among the lowest (~7328 for F, ~7415 for M). The differences between genders within each education level are relatively small. This suggests education level alone is not a strong predictor of higher customer value.

## Bonus

You work at the customer service department and you want to know which months had the highest number of complaints by policy type category. Create a summary table showing the number of complaints by policy type and month.
Show it in a long format table.

*In data analysis, a long format table is a way of structuring data in which each observation or measurement is stored in a separate row of the table. The key characteristic of a long format table is that each column represents a single variable, and each row represents a single observation of that variable.*

*More information about long and wide format tables here: https://www.statology.org/long-vs-wide-data/*

In [29]:
# Your code goes here